File: Solution Customized Linked List.cpp

#include <memory>
#include <string>
using namespace std;

template<typename T> struct Website {
    shared_ptr<Website<T>> forward;
    shared_ptr<Website<T>> backward;
    T address;
    
    explicit Website(T address) : address{move(address)}{}
};

class BrowserHistory {
    
    shared_ptr<Website<string>> currentWebsite;

public:
    explicit BrowserHistory(string homepage): currentWebsite {make_shared<Website<string>>(move(homepage))}{}

    void visit(string url) {
        auto newWebsite = make_shared<Website<string>>(move(url));
        currentWebsite->forward = newWebsite;
        newWebsite->backward = currentWebsite;
        currentWebsite = newWebsite;
    }

    string back(int steps) {
        while (steps-- > 0 && currentWebsite->backward != nullptr) {
            currentWebsite = currentWebsite->backward;
        }
        return currentWebsite->address;
    }

    string forward(int steps) {
        while (steps-- > 0 && currentWebsite->forward != nullptr) {
            currentWebsite = currentWebsite->forward;
        }
        return currentWebsite->address;
    }
};

File: Solution Customized Linked List.js

class BrowserHistory {

    /**
     * @param {string} homepage
     */
    constructor(homepage) {
        this.currentWebsite = new Website(homepage);
    }

    /** 
     * @param {string} url
     * @return {void}
     */
    visit(url) {
        let newWebsite = new Website(url);
        this.currentWebsite.forward = newWebsite;
        newWebsite.backward = this.currentWebsite;
        this.currentWebsite = newWebsite;
    }

    /** 
     * @param {number} steps
     * @return {string}
     */
    back(steps) {
        while (steps-- > 0 && this.currentWebsite.backward !== null) {
            this.currentWebsite = this.currentWebsite.backward;
        }
        return this.currentWebsite.address;
    }

    /** 
     * @param {number} steps
     * @return {string}
     */
    forward(steps) {
        while (steps-- > 0 && this.currentWebsite.forward !== null) {
            this.currentWebsite = this.currentWebsite.forward;
        }
        return this.currentWebsite.address;
    }
}

class Website {

    /**
     * @param {string} address
     */
    constructor(address) {
        this.forward = null;
        this.backward = null;
        this.address = address;
    }
}
